geps/gep-1619.md

```diff
@@ -163,43 +163,299 @@ Java application servers such as Tomcat and Jetty, were the first to standardize
+configuration for every route that utilizes the service.
+
+Conversely, when attaching session persistence configuration to xRoutes, it becomes possible to establish a clear
+relationship that it augments routing behavior, particularly in scenarios involving traffic splitting. However, as
+mentioned earlier, session persistence would need to be individually configured for each specific route.
+
+**How are existing implementation APIs designed?**
+
+Referring to our [Implementations](#Implementations) tables on session persistence, the majority of Gateway API
+implementations designed session persistence in their APIs to be attached to a service or backends. This should be
+considered cautiously, as making associations to Gateway API's notion of Gateway, Route, and Service to other
+implementation's objects is hard to directly translate. The idea of a route in Gateway API is often not the same as a
+route in any given implementation.
+
+Taking Istio as an example, it offers an attachment point of session persistence to a service through a [`DestinationRule`](https://istio.io/latest/docs/reference/config/networking/destination-rule/#DestinationRule).
+`DestinationRule` defines policies that apply to traffic intended for a service after routing has occurred. Within
+`DestinationRule`, Istio couples the selection of load balancing algorithms with session persistence via their [`LoadBalancerSettings`](https://istio.io/latest/docs/reference/config/networking/destination-rule/#LoadBalancerSettings).
+When designing our API structure, it is important to consider the balance between incorporating additional load
+balancing policies coupled with session persistence, rather than configuring session persistence independently.
+
+On the other hand, rather than being attached to a service, Envoy configures session persistence at the route-level
+through an [`HttpFilter`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#extensions-filters-network-http-connection-manager-v3-httpfilter).
+As previously mentioned, this approach allows data plane entities to implement session persistence across services
+(clusters), thereby augmenting routing behavior in scenarios involving traffic splitting.
+
+#### Session Persistence API with GAMMA
+
+The object of the [GAMMA (Gateway API for Mesh Management and Administration)](https://gateway-api.sigs.k8s.io/contributing/gamma/)
+initiative is to provide support for service mesh and mesh-adjacent use-cases with Gateway API. Mesh traffic routing with
+Mesh traffic routing with Gateway API is accomplished with `xRoutes` specifying service-to-service traffic opposed to
+gateway-to-service traffic.
+
+Given that service meshes commonly have session persistence requirements, this API design should take into consideration
+session persistence needs in GAMMA and service mesh scenarios.
+
+#### API Design
```

Comment:
Overall the API itself seems reasonable. One thing that isn't clear to me is how its meant to be implemented.

We define all the attributes of the cookie, but what is the _value_ of the cookie? is it an implementation specific opaque value?

In Envoy's currently implementation it is an ip+port pair encoded as base64 I believe. This approach is somewhat flawed when paired with other gateway features like weighted traffic routes. Consider we have backends with IPs [a,b] and [c,d]. We have persistent sessions configure for the route and 50/50 split.

Our first request would pick a backend and specific IP, say `a`. Then the next request would pick the other backend (50/50 chance); there is no `a` there so persistence is lost.

Is this in incorrect implementation of the spec? Is it open-ended on behavior here?

Reply:
@howardjohn I had raised this point earlier in https://github.com/kubernetes-sigs/gateway-api/pull/2159#discussion_r1252413412 and it seems @gcs278 has acknowledged and addressed it (haven't had time to read the responses yet).

In terms of cookie value, we need to add the cluster (aka Service) name as described in https://github.com/grpc/proposal/blob/master/A60-xds-stateful-session-affinity-weighted-clusters.md#include-cluster-name-in-the-cookie .

Reply:
I agree with @costinm that the cookie value must be opaque and left to the implementations to decide. Of course the spec here will dictate what minimum information the cookie should contain.